### PR TITLE
Small improvements found by my phpstan work

### PR DESCRIPTION
**Pull request by sminnee**
This is a trimming back of a rebased version of PHPStan PR, #9103. 

The work of preparing that led me to find lots of little things that could be tidied up that, while they would require PR, are safe changes.

I've moved all the 'dangerous' changes out of that PR but figure it's useful to get these into mainline so that they are less likely to rot.

In particular, I note that @sunnysideup has started some PHPStan work in parallel and it's a bit of a shame that this work was rotting on a branch.

The original PR still has 5 commits not included in this.

### CODE REVIEW

**Comment by szepeviktor** [comment]
@Cheddam Could you help us?

File: src/Control/HTTPRequest.php [hunk 1]
```diff
@@ -503,7 +503,7 @@ public static function send_file($fileData, $fileName, $mimeType = null)
      * The pattern can optionally start with an HTTP method and a space.  For example, "POST $Controller/$Action".
      * This is used to define a rule that only matches on a specific HTTP method.
      *
-     * @param $pattern
+     * @param string$pattern
```
Comment (szepeviktor) [inline src/Control/HTTPRequest.php:506]: Here we need one more whitespace.

File: src/Forms/ConfirmedPasswordField.php [hunk 2]
```diff
@@ -136,6 +136,9 @@ public function __construct(
         $titleConfirmField = null
     ) {
 
+        // This param is requested only for backward compat
+        $form;
```
Comment (szepeviktor) [inline src/Forms/ConfirmedPasswordField.php:140]: Maybe `ignoreErrors:` is a better place than this noop.

File: src/Forms/GridField/GridFieldDeleteAction.php [hunk 3]
```diff
@@ -168,6 +170,25 @@ public function getColumnContent($gridField, $record, $columnName)
         return null;
     }
 
+    /**
+     * Check that this dataList is of the right data type.
+     * Returns false if it's a bad data type, and if appropriate, throws an exception.
+     *
+     * @param SS_List $dataList
+     * @return bool
```
Comment (szepeviktor) [inline src/Forms/GridField/GridFieldDeleteAction.php:178]: This could be `true`.

File: src/Forms/GridField/GridFieldDeleteAction.php [hunk 4]
```diff
@@ -168,6 +170,25 @@ public function getColumnContent($gridField, $record, $columnName)
...
+     */
+    protected function checkDataType($dataList)
+    {
+        if ($dataList instanceof Filterable) {
+            return true;
+
+        } else {
```
Comment (szepeviktor) [inline src/Forms/GridField/GridFieldDeleteAction.php:185]: No need for `else`.

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php [hunk 5]
```diff
@@ -808,4 +813,22 @@ public function Breadcrumbs($unlinked = false)
...
+     * Throws an exception if the data type is bad
+     *
+     * @param SS_List $dataList
+     * @return SS_List&Filterable&Limitable
+     */
+    protected function checkDataType($dataList)
+    {
+        if ($dataList instanceof Filterable && $dataList instanceof Limitable) {
+            return $dataList;
+        } else {
```
Comment (szepeviktor) [inline src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:828]: No need for `else`.

File: src/Forms/SingleSelectField.php [hunk 6]
```diff
@@ -102,13 +103,17 @@ public function getEmptyString()
...
     {
         // Inject default option
         if ($this->getHasEmptyDefault()) {
-            return ['' => $this->getEmptyString()] + $this->getSource();
+            $source = $this->getSource();
+            if (!is_array($source)) {
+                $source = iterator_to_array($source, true);
+            }
+            return ['' => $this->getEmptyString()] + $source;
         } else {
```
Comment (szepeviktor) [inline src/Forms/SingleSelectField.php:117]: No need for `else`.

File: src/ORM/DatabaseAdmin.php [hunk 7]
```diff
@@ -322,7 +322,7 @@ public function doBuild($quiet = false, $populate = true, $testMode = false)
                         try {
                             $count = DB::query("SELECT COUNT(*) FROM \"$tableName\"")->value();
                             $countSuffix = " ($count records)";
-                        } catch (Exception $e) {
+                        } catch (\Exception $e) {
```
Comment (szepeviktor) [inline src/ORM/DatabaseAdmin.php:325]: Please import that `use Exception;`

File: src/ORM/ListDecorator.php [hunk 8]
```diff
@@ -99,7 +99,11 @@ public function getIterator()
 
     public function exists()
     {
-        return $this->list->exists();
+        if (method_exists($this->list, 'exists')) {
+            return $this->list->exists();
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:104]: No need for `else`.

File: src/ORM/ListDecorator.php [hunk 9]
```diff
@@ -124,7 +128,11 @@ public function Count()
 
     public function forTemplate()
     {
-        return $this->list->forTemplate();
+        if (method_exists($this->list, 'forTemplate')) {
+            return $this->list->forTemplate();
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:133]: else!

File: src/ORM/ListDecorator.php [hunk 10]
```diff
@@ -144,7 +152,11 @@ public function column($value = 'ID')
 
     public function columnUnique($value = "ID")
     {
-        return $this->list->columnUnique($value);
+        if (method_exists($this->list, 'columnUnique')) {
+            return $this->list->columnUnique($value);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:157]: else!

File: src/ORM/ListDecorator.php [hunk 11]
```diff
@@ -154,12 +166,20 @@ public function each($callback)
 
     public function canSortBy($by)
     {
-        return $this->list->canSortBy($by);
+        if ($this->list instanceof Sortable) {
+            return $this->list->canSortBy($by);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:171]: else!

File: src/ORM/ListDecorator.php [hunk 12]
```diff
@@ -154,12 +166,20 @@ public function each($callback)
...
+            throw new \BadMethodCallException('canSortBy() not found; the decorated list isn\'t Sortable');
+        }
     }
 
     public function reverse()
     {
-        return $this->list->reverse();
+        if ($this->list instanceof Sortable) {
+            return $this->list->reverse();
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:180]: else!

File: src/ORM/ListDecorator.php [hunk 13]
```diff
@@ -171,14 +191,22 @@ public function reverse()
      * @example $list->sort('Name', 'ASC');
      * @example $list->sort(array('Name'=>'ASC,'Age'=>'DESC'));
      */
-    public function sort()
+    public function sort(...$arguments)
     {
-        return $this->list->sort(...func_get_args());
+        if ($this->list instanceof Sortable) {
+            return $this->list->sort(...$arguments);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:198]: else!

File: src/ORM/ListDecorator.php [hunk 14]
```diff
@@ -171,14 +191,22 @@ public function reverse()
...
+            throw new \BadMethodCallException('sort() not found; the decorated list isn\'t Sortable');
+        }
     }
 
     public function canFilterBy($by)
     {
-        return $this->list->canFilterBy($by);
+        if ($this->list instanceof Filterable) {
+            return $this->list->canFilterBy($by);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:207]: else!

File: src/ORM/ListDecorator.php [hunk 15]
```diff
@@ -189,9 +217,13 @@ public function canFilterBy($by)
      * @example $list->filter(array('Name'=>'bob, 'Age'=>21)); // bob or someone with Age 21
      * @example $list->filter(array('Name'=>'bob, 'Age'=>array(21, 43))); // bob or anyone with Age 21 or 43
      */
-    public function filter()
+    public function filter(...$arguments)
     {
-        return $this->list->filter(...func_get_args());
+        if ($this->list instanceof Filterable) {
+            return $this->list->filter(...$arguments);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:224]: else!

File: src/ORM/ListDecorator.php [hunk 16]
```diff
@@ -213,12 +245,16 @@ public function filter()
...
+     * @param array<int,mixed> $arguments See {@link filter()}
      * @return DataList
      */
-    public function filterAny()
+    public function filterAny(...$arguments)
     {
-        return $this->list->filterAny(...func_get_args());
+        if ($this->list instanceof Filterable) {
+            return $this->list->filterAny(...$arguments);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:255]: else!

File: src/ORM/ListDecorator.php [hunk 17]
```diff
@@ -249,7 +285,11 @@ public function filterByCallback($callback)
 
     public function limit($limit, $offset = 0)
     {
-        return $this->list->limit($limit, $offset);
+        if ($this->list instanceof Limitable) {
+            return $this->list->limit($limit, $offset);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:290]: else!

File: src/ORM/ListDecorator.php [hunk 18]
```diff
@@ -260,7 +300,11 @@ public function limit($limit, $offset = 0)
      */
     public function byID($id)
     {
-        return $this->list->byID($id);
+        if ($this->list instanceof Filterable) {
+            return $this->list->byID($id);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:305]: else!

File: src/ORM/ListDecorator.php [hunk 19]
```diff
@@ -271,7 +315,11 @@ public function byID($id)
      */
     public function byIDs($ids)
     {
-        return $this->list->byIDs($ids);
+        if ($this->list instanceof Filterable) {
+            return $this->list->byIDs($ids);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:320]: else!

File: src/ORM/ListDecorator.php [hunk 20]
```diff
@@ -282,13 +330,21 @@ public function byIDs($ids)
      * @example $list->exclude(array('Name'=>'bob, 'Age'=>21)); // exclude bob or someone with Age 21
      * @example $list->exclude(array('Name'=>'bob, 'Age'=>array(21, 43))); // exclude bob or anyone with Age 21 or 43
      */
-    public function exclude()
+    public function exclude(...$arguments)
     {
-        return $this->list->exclude(...func_get_args());
+        if ($this->list instanceof Filterable) {
+            return $this->list->exclude(...$arguments);
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:337]: else!

File: src/ORM/ListDecorator.php [hunk 21]
```diff
@@ -282,13 +330,21 @@ public function byIDs($ids)
...
+            throw new \BadMethodCallException('exclude() not found; the decorated list isn\'t Filterable');
+        }
     }
 
     public function debug()
     {
-        return $this->list->debug();
+        if ($this->list instanceof ViewableData) {
+            return $this->list->debug();
+        } else {
```
Comment (szepeviktor) [inline src/ORM/ListDecorator.php:346]: else!

**Review by szepeviktor — APPROVED** [review]

**Review by szepeviktor — CHANGES_REQUESTED** [review]
Please consider putting errors in conditional and normal execution without indentation.

```php
if (we have a problem) {
    // Error here.
    throw new Exception('💣');
}

return $normal . $execution;
```

Normal execution in a conditional is hard to read: zig-zag instead of straight down.

**Review by chillu — CHANGES_REQUESTED** [review]
Looked through all of this - again, what a mission, well done Sam! Apart from the three review comments I'm happy for this to go in.

File: src/ORM/Connect/MySQLDatabase.php [hunk 22]
```diff
@@ -368,20 +368,6 @@ public function transactionEnd($chain = false)
         return $result;
     }
 
-    /**
-     * In error condition, set transactionNesting to zero
-     */
-    protected function resetTransactionNesting()
```
[the remark this answers is not in the document]
Comment (sminnee) [reply src/ORM/Connect/MySQLDatabase.php:373]: Right I'll pull the change out of this PR and into the other one for more in-depth discussion

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php [hunk 23]
```diff
@@ -806,4 +813,22 @@ public function Breadcrumbs($unlinked = false)
...
+     * Check that this dataList is of the right data type.
+     * Throws an exception if the data type is bad
+     *
+     * @param SS_List $dataList
+     * @return SS_List&Filterable&Limitable
+     */
+    protected function checkDataType($dataList)
+    {
+        if ($dataList instanceof Filterable && $dataList instanceof Limitable) {
+            return $dataList;
```
[the remark this answers is not in the document]
Comment (sminnee) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:827]: The point was to return something of a known type but fair enough an @var works too.

I might try an assert() or something as well...

File: src/ORM/DataObject.php [hunk 24]
```diff
@@ -1146,13 +1146,11 @@ public function merge($rightObj, $priority = 'right', $includeRelations = true,
         if ($includeRelations) {
             if ($manyMany = $this->manyMany()) {
                 foreach ($manyMany as $relationship => $class) {
-                    /** @var DataObject $leftComponents */
                     $leftComponents = $leftObj->getManyManyComponents($relationship);
                     $rightComponents = $rightObj->getManyManyComponents($relationship);
                     if ($rightComponents && $rightComponents->exists()) {
                         $leftComponents->addMany($rightComponents->column('ID'));
                     }
-                    $leftComponents->write();
```
[the remark this answers is not in the document]
Comment (sminnee) [reply src/ORM/DataObject.php:1155]: Yeah I think so. This was something that was identified by failures in type resolution, and it was a while ago. It probably deserves a unit test.

File: src/Forms/ConfirmedPasswordField.php (hunk 2 above)
Comment (sminnee) [reply src/Forms/ConfirmedPasswordField.php:140]: I like this better.

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php (hunk 23 above)
[the remark this answers is not in the document]
Comment (sminnee) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:827]: OK so `assert()`s work with PHPStan: https://phpstan.org/r/9b405a35-5203-45f7-9cd0-2e611d9e2331 

I'm checking PHPStorm now too.

[the remark this answers is not in the document]
Comment (sminnee) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:827]: PHPStorm and VS Code both use `assert(... instanceof ...)` commands to augment type-hinting.

I'll use that.

[the remark this answers is not in the document]
Comment (sminnee) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:827]: I'd note that `/** @var */` is the worst possible solution here as it doesn't affect code execution. It just helps your IDE lie to you. :-P

**Pull request closed without merging**